### ISSUE TITLE
Add ability to specify origin certificate secrets as CLI flag or env var 

### ISSUE THREAD

**Issue by hankjacobs** [cloudflare/cloudflare-ingress-controller#126]
Prior to v0.6.0, domain certificate secrets were found using labels on the secret. In v0.6.0, this was removed in favor of the tls section of each ingress. For our use case, this is a regression. We will have many, many ingresses (in the hundreds) spanning many namespaces (~80 or 90) and adding secrets to each namespace and modifying each ingress is a large amount of work and is hard to maintain as we add more domains. 

In the current version, I see one can set a default origin secret via a CLI flag. It would be great to add the ability to specify multiple origin secrets via a CLI flag or env var. This would totally fix our issue with the way origin secrets are handled in v0.6.0.

**Comment by mattalberts** [comment]
@hankjacobs for clarity, 
> We will have many, many ingresses (in the hundreds) spanning many namespaces (~80 or 90) and adding secrets to each namespace and modifying each ingress is a large amount of work and is hard to maintain as we add more domains

- "will have many" - Do you have then now? Or will you have them in the future? (I also operate in an environment with many ingress definitions, making mass editing hairy)
- "Adding secrets to each namespace" - what does the overlap look like here (especially in context of the PR), do tunnels/ingresses in different namespaces share different secrets? That's a hard one to phrase, but if we call the default an N-to-1 relationship, it sounds like you are describing an N-to-M relationship for secrets, where ingresses in other namespaces make use of the same secret, beyond the default secret.  Is that true? Are you not concerned about access to all the secrets?

I want to make certain that I understand the ask and the issue directly, especially given how large of a change in behavior the old version was from all other ingress controllers. In a multi-tenant environment, managing the secrets next to the ingress means user only ever needs to have access to the secret for they service they operate (and can update that secret when it expires directly).  It also allows users to add new services and secrets without needing access to the central location.

**Comment by hankjacobs** [comment]
- We are in the process of rolling out an update to an internal system (more details in answer below) that makes use of Argo ingresses. Currently, only a fraction of the namespaces use Argo ingresses but this increases on a daily basis.
- I _think_ I understand the question but feel free to correct me if I'm off base. It's an N-to-M relationship. [This blog post](https://engineering.dollarshaveclub.com/qa-environments-on-demand-with-kubernetes-5a571b4e273c) goes into details about the system we are upgrading to use argo ingresses. The TLDR is that each namespace has all the resources needed to run a QA environment for applications we develop. Each namespace has a set of ingresses. Each ingress within a namespace is a subdomain on one of 6 (maybe more later) argo tunnel domains (the ingresses are usually all on the same domain but not necessarily). Users have access to these QA namespaces but do not have access to the namespace that the ingress controller lives in. To support the way v0.6.0 does ingress secrets, we'd have to either a) install all the secrets for the available ingress domains into the QA namespace (which would make the secrets available to the user) or b) determine at runtime which domains were selected and install just that one (which still makes the secret available to the user). Access to the secrets is actually more of a concern with v0.6.0 since the secrets would be in a namespace users have access to vs v0.5.0 where they are in an entirely different namespace they don't have access to. Additionally, these QA namespaces are short lived (think 1 hour or less).

To speak to your point on a multi-tenant environment, we consider the ingress controllers a service our team provides our engineering teams. This includes provisioning new Argo tunnel domains, rotating certs, etc etc. Most developers do not have access to Cloudflare so they couldn't do any of the above if they wanted to. 

Let me know if you'd like me to clarify any of the above points. Thanks!

**Comment by mattalberts** [comment]
That totally helps! Thank you!

**Comment by mattalberts** [comment]
@hankjacobs I ended up backing off from introducing a config map. To get the proper rebuild behavior, I'd need to build an index by ingress host, which felt like too much additional complication.  Instead, I've merged our two ideas into a command-line flag used to specify a configuration file that maps hosts to secrets to provide host-specific defaults (which will allow the an Ingress to provide a different secret if desirable, but also allows you to locate all the reusable secrets in an isolated namespace). Hows this look?

**Comment by hankjacobs** [comment]
Sounds reasonable. That would work for our use-case. Thanks!

**Comment by mattalberts** [comment]
@hankjacobs the update was release with 0.6.2.  Would love to if it works out for you or if we need to go back to the drawing board a bit

**Comment by hankjacobs** [comment]
@mattalberts Thanks for the update! The design looks great. Using a config map to store the mappings made it easy to implement. 

That being said, we had to rollback. It appears that you are required to specify the full hostname used by the tunnel within the origin-secret-config. In 0.5.x, the secret only had to specify the domain (or subdomain) upon which the tunnel is created. (e.g `google.com` for any tunnel that matches `*.google.com`). Our tunnel domains are generated on the fly so we don't know the list up front and we'd rather not have to update the config map and reload argo every time a tunnel is added or removed. 

Ideally, the origin-secret-config is updated to allow wildcards. Just my two cents though.

Let us know what you think and thanks for iterating on this with us!

**Comment by mattalberts** [comment]
Which version did you roll-back to? 0.5.3?

**Comment by mattalberts** [comment]
Yeah, I was very worried about people providing both fully-qualified hosts and wildcards and then not understanding that based on order, one might cover the other. There are ways to work around that however.

**Comment by hankjacobs** [comment]
Totally makes sense. Maybe our use-case is unconventional but we're big fans of argo-tunnel and the way v0.5.3 behaves because it has greatly simplified one of our internal systems (that was Route53 + ELB based prior to Argo).

**Comment by mattalberts** [comment]
I have a few ideas, I think I'll put together a PR and send it to you to review, just to make certain it covers your use-case.

**Comment by hankjacobs** [comment]
:+1: excellent. looking forward to it. thanks.

**Comment by mattalberts** [comment]
@hankjacobs Okay, check this out.  I think it gets you where you need to be in terms of controls. If not, I have one other crazy idea :)

**Comment by hankjacobs** [comment]
That looks like it'll work for us. I'll test it out as soon as you cut a release. Thanks!

**Comment by mattalberts** [comment]
this feels like a good progression forward.  The next generation of this will allow secret-namespace overrride as an annotation (that will get things to be more dynamic).

**Comment by bkeroackdsc** [comment]
I just want to leave a note that we tried to upgrade to 0.6.3 and ran into a critical bug, forcing us to roll back to 0.5.x again.

**Comment by mattalberts** [comment]
Benjamen,

Did you try upgrading to the custom patch build?

**Comment by mattalberts** [comment]
@bkeroackdsc I've bundled the changes from the patch build with additional updates, please try v0.6.5
 - https://github.com/cloudflare/cloudflare-ingress-controller/releases/tag/0.6.5
